src/net.rs

```diff
@@ -0,0 +1,168 @@
+/// physical layer's packets.
+///
+/// A `HostPort` provides all low-level management of incomming and outgoing
+/// connections. An implementation is responsible for packet assembly,
+/// verifying redundancy codes, tracking AEAD tags, and so on, exposing only
+/// an abstract header and a I/O for the message payload itself to Manticore.
+///
+/// This trait is object-safe.
+///
+/// # Example
+/// ```
+/// # use manticore::mem::*;
+/// # use manticore::net::*;
+/// # use manticore::protocol::*;
+/// # use manticore::protocol::firmware_version::*;
+/// # use manticore::protocol::wire::*;
+/// # let mut rx = [0; 32];
+/// # let mut tx = [0; 32];
+/// # let mut port = InMemPort {
+/// #     addr: Address([1, 2, 3, 4]),
+/// #     rx_header: Some(Header {
+/// #         command: CommandType::FirmwareVersion,
+/// #         is_request: true,
+/// #     }),
+/// #     rx: &mut rx,
+/// #     tx_header: None,
+/// #     tx: &mut tx,
+/// # };
+/// # let mut arena = [0; 64];
+/// # let arena = BumpArena::new(&mut arena);
+/// let (addr, header, read) = port.receive()?;
+/// assert_eq!(header.command, CommandType::FirmwareVersion);
+/// assert!(header.is_request);
+///
+/// let req = FirmwareVersionRequest::from_wire(read, &arena).unwrap();
+/// assert_eq!(req.index, 0);
+///
+/// let resp = FirmwareVersionResponse {
+///     version: &[0xba; 32],
+/// };
+/// let write = port.reply(addr, Header {
+///     command: CommandType::FirmwareVersion,
+///     is_request: false,
+/// })?;
+/// resp.to_wire(write).unwrap();
+/// // write.flush()?;
+/// # Ok::<(), manticore::net::Error>(())
+/// ```
+pub trait HostPort {
+    /// Receives an incomming message from a connected host device.
+    ///
+    /// This function will block until a host device indicates that it wishes
+    /// to communicate. At that point, this function will perform enough
+    /// transport-level operations to populate an abstract Manticore header,
+    /// and then hand off parsing to the caller.
+    ///
+    /// Upon success, this function will return:
+    /// - An abstract address, representing the overal request (used with
+    ///   `reply()` to indicate which host to reply to).
+    /// - A populated Manticore header.
+    /// - A [`Read`], which can be used to consume the payload of the message.
+    ///   this [`Read`] *must not* make any other bytes acessible.
+    ///
+    /// [`Read`]: ../io/trait.Read.html
+    fn receive(&mut self) -> Result<(Address, Header, &mut dyn Read), Error>;
+
+    /// Replies to the request indicated by `addr`.
+    ///
+    /// Calling this function performs sufficient transport-level operations to
+    /// begin a response, before handing off actually composing the payload to
+    /// the caller via a [`Write`].
+    ///
+    /// `addr` must have been previously returned by `recieve()`. Using any
+    /// other value of `addr`, or using it repeatedly, is unspecified behavior.
```

Comment:
Must the `addr` passed to `reply()` always be the `Address` returned by the most recent call to `receive()`? Or can it be from an older call to `receive()`, so long as that `Address` has not yet been used in a reply? (sorry, I am not familiar with the protocol).

If the former, does it make more sense to leave implementations responsible for storing and reusing the `Address`, rather than leaving it to users of the interface where mistakes are perhaps more likely? For the `InMemPort` example it seems like having reply always use `self.addr` is preferable to requiring the caller to pass the address to `reply()`.

Reply:
I actually think you're right. The reason for having `Address` was so that a `HostPort` could have multiple outstanding requests... but in all the likely uses of Manticore, that's probably not going to happen. An eventual `DevicePort` trait, however, will likely need some kind of addressing mechanism.

> The pattern of having the actual payload not be passed to reply() feels a little weird also, but I am guessing there is a reason for letting the user of the interface manually interact with the Write object?

This is to minimize the amount of buffering needed. Rather than requiring the caller to serialize their message somewhere and *then* pipe it into the port, the port is responsible for doing all of that in the most efficient manner possible.